ouroboros-network/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -0,0 +1,265 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+
+module Ouroboros.Network.Subscription.Dns
+    ( DnsSubscriptionTarget (..)
+    , Resolver (..)
+    , dnsSubscriptionWorker'
+    , dnsSubscriptionWorker
+    , dnsResolve
+    , resolutionDelay
+    ) where
+
+import           Control.Monad (unless)
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSay
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadThrow
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Tracer
+import qualified Data.IP as IP
+import qualified Network.DNS as DNS
+import qualified Network.Socket as Socket
+import           Text.Printf
+
+import           Ouroboros.Network.Subscription.Common
+import           Ouroboros.Network.Subscription.Subscriber
+import           Ouroboros.Network.Socket
+
+-- | Time to wait for an AAAA response after receiving an A response.
+resolutionDelay :: DiffTime
+resolutionDelay = 0.05 -- 50ms delay
+
+data DnsSubscriptionTarget = DnsSubscriptionTarget {
+      dstDomain :: !DNS.Domain
+    , dstPort   :: !Socket.PortNumber
+    , dstValency :: !Int
+    } deriving (Eq, Show)
+
+data Resolver m = Resolver {
+      lookupA    :: DNS.Domain -> m (Either DNS.DNSError [Socket.SockAddr])
+    , lookupAAAA :: DNS.Domain -> m (Either DNS.DNSError [Socket.SockAddr])
+    }
+
+dnsResolve :: forall m.
+     ( MonadAsync m
+     , MonadSay   m
+     , MonadSTM   m
+     , MonadTime  m
+     , MonadTimer m
+     , MonadThrow m
+     )
+    => Tracer m DnsTrace
+    -> Resolver m
+    -> DnsSubscriptionTarget
+    -> m (SubscriptionTarget m Socket.SockAddr)
+dnsResolve tracer resolver (DnsSubscriptionTarget domain _ _) = do
+    ipv6Rsps <- newEmptyTMVarM
+    ipv4Rsps <- newEmptyTMVarM
+    gotIpv6Rsp <- newTVarM False
+
+    aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
+    aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
+
+    rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
+    case rd_e of
+         Left r_ipv6 -> -- AAAA lookup finished first
+             case r_ipv6 of
+                  Left e_ipv6 -> do -- AAAA lookup failed
```

Comment:
We could combine this into a nested pattern, rather than nested case exprs. Then we don't have to invent extra var names like `r_ipv6`.

Reply:
Will address.